walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,178 @@
 
 package com.google.sps;
 
+import java.util.*;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Meeting is longer than the whole day, return an empty array.
+    if (request.getDuration() >= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no events, return the whole day as time range.
+    if (events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> result;
+    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> requestAttendeesOptional = request.getOptionalAttendees();
+
+    // If there are no attendees, return the whole day.
+    if (requestAttendees.isEmpty() && requestAttendeesOptional.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Get unavailable times for mandatory attendees.
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, requestAttendees);
+    Collections.sort(unavailableTimes, TimeRange.ORDER_BY_START);
+
+    // Get unavailable times for optional attendees.
+    List<TimeRange> unavailableTimesOptional = getUnavailableTimes(events, requestAttendeesOptional);
+    Collections.sort(unavailableTimesOptional, TimeRange.ORDER_BY_START);
+
+    // If there are no mandatory attendees, optional attendees are now mandatory and unavailableTimesOptional becomes empty
+    if (requestAttendees.isEmpty()) {
+      unavailableTimes = unavailableTimesOptional;
+      unavailableTimesOptional = new ArrayList<>();
+    }
+
+    return getAvailableTimes(request.getDuration(), unavailableTimes, unavailableTimesOptional);
+
   }
+
+  /**
+   * Get all unavailable times in which meeting request attendees are busy.
+   * 
+   * @param events to get unavailabilities from
+   * @param requestAttendees to find overlap for
+   * @return list of time ranges that are already occupied
+   */
+  public List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> requestAttendees) {
+    List<TimeRange> output = new ArrayList<>();
```

Comment:
I'm curious if you want to use a Set here instead of List, if you don't care about duplicates.

Reply:
I wanted to use a list because I could sort it for the mergeOverlappingTimes function such that I wouldn't have to use two for loops!!